clients/adapter-amqp/src/main/java/org/eclipse/hono/adapter/client/amqp/RequestResponseClient.java

```diff
@@ -0,0 +1,647 @@
+                                requestTargetAddress, request.getSubject(), correlationId, remoteState);
+                        final Modified modified = (Modified) deliveryUpdated.getRemoteState();
+                        failedResult.fail(modified.getUndeliverableHere() ? new ClientErrorException(HttpURLConnection.HTTP_NOT_FOUND)
+                                : new ServerErrorException(HttpURLConnection.HTTP_UNAVAILABLE));
+                        cancelRequest(correlationId, failedResult.future());
+                    } else if (remoteState == null) {
+                        // possible scenario here: sender link got closed while waiting on the delivery update
+                        final String furtherInfo = !sender.isOpen() ? ", sender link was closed in between" : "";
+                        LOG.warn("got undefined delivery state for service request{} [target address: {}, subject: {}, correlation ID: {}]",
+                                furtherInfo, requestTargetAddress, request.getSubject(), correlationId);
+                        failedResult.fail(new ServerErrorException(HttpURLConnection.HTTP_UNAVAILABLE));
+                        cancelRequest(correlationId, failedResult.future());
+                    }
+                });
+                if (requestTimeoutMillis > 0) {
+                    connection.getVertx().setTimer(requestTimeoutMillis, tid -> {
+                        cancelRequest(correlationId, Future.failedFuture(new ServerErrorException(
+                                HttpURLConnection.HTTP_UNAVAILABLE, "request timed out after " + requestTimeoutMillis + "ms")));
+                        sample.timeout();
+                    });
+                }
+                if (LOG.isDebugEnabled()) {
+                    final String deviceId = MessageHelper.getDeviceId(request);
+                    if (deviceId == null) {
+                        LOG.debug("sent request [target address: {}, subject: {}, correlation ID: {}] to service",
+                                requestTargetAddress, request.getSubject(), correlationId);
+                    } else {
+                        LOG.debug("sent request [target address: {}, subject: {}, correlation ID: {}, device ID: {}] to service",
+                                requestTargetAddress, request.getSubject(), correlationId, deviceId);
+                    }
+                }
+            }
+          });
+    }
+
+    /**
+     * Checks if this client's sender and receiver links are open.
+     *
+     * @return {@code true} if a request can be sent to and a response can be received
+     * from the peer.
+     */
+    public final boolean isOpen() {
+        return sender != null && sender.isOpen() && receiver != null && receiver.isOpen();
+    }
+
+    /**
+     * Closes the links.
+     *
+     * @param closeHandler The handler to invoke with the result.
+     */
+    public final void close(final Handler<AsyncResult<Void>> closeHandler) {
```

Comment:
Could be refactored to return `Future<Void>` instead of taking the handler parameter (which would be in line with `createLinks` for example).